projects/project5/src/main/scala/TransH.scala

```diff
@@ -0,0 +1,391 @@
+extends Models(ne: Int, nr: Int, batch: Int, k: Int, sk: SparkSession){
+  
+  //Calculate the number of batches to be run in an epoch based on minibatch size and size of training data 
+  val nBatches = (train.count() / batch).toInt
+  
+  //Initialize Tensor containing normal vectors w for each relation
+  var wR: Tensor[Float] = normalizeTensor(initialize(nr), true)
+  
+  //We tried to use Adam Method from bigdl library to learn our parameters but we faced difficulties
+  //by the way it was constructed, the adam method was a parameter of a Module Class and creating a custom module
+  //was strongly coupled to a neural network approach which was not compatible for our TransH model
+  //var opt = new Adam(learningRate = rate)
+  
+  def run(): Unit = {
+
+    for (i <- 1 to epochs) {
+      
+      println("Start epoch " + i)
+      //Accumulator to store the total loss
+      var ac_res = sk.sparkContext.longAccumulator
+      //Starting time of an epoch
+      var start = System.currentTimeMillis()
+
+      //Loop 1 to number of batches for an epoch
+      for(batch <- 1 to 50){
+        
+        //Broadcast the embeddings of entities, relations and normal vectors w
+        val bcE = sk.sparkContext.broadcast(e)
+        val bcR = sk.sparkContext.broadcast(r)
+        val bcWr = sk.sparkContext.broadcast(wR)
+        
+//        if(batch%10 == 0)
+//          println(batch)
+        
+        //Get a sample/minibatch of the training data
+        val sample = subset(train).rdd
+        
+        /**
+         * We used map partitions because we found it better suited for this case and also it's faster than map function.
+         * For each partition of the RDD containing the sample data we run the TransH learning algorithm:
+         * 1) For each tuple corrupt the head and tail (one at a time) by a random entity in the knowledge graph
+         * 2) Calculate the loss/error from a true tuple and the corresponding corrupted tuple
+         * 3) If the loss/error is greater than 0 then optimize the parameters (entity, relation, normal vector w embeddings)
+         * 4) After optimizing for a tuple, make sure the TransH model constraints hold:
+         * 		- L2 norm of an embedding of an entity <= 1
+         * 		- the embeddings of entities must be orthogonal to the normal vectors w (dot product = 0 theoretically)
+         * 		- L2 norm of the normal vectors w = 1
+         */
+        val partitions = sample.mapPartitions{
+          iterator =>
+            
+            //Create maps to contain the embeddings of entities, relations, normal vectors w in each partition
+            var ePerPartition = Map[Int,Tensor[Float]]()
+            var rPerPartition = Map[Int,Tensor[Float]]()
+            var wrPerPartition = Map[Int,Tensor[Float]]()
+            var retMap = ArrayBuffer[(Map[Int,Tensor[Float]],Map[Int,Tensor[Float]], Map[Int,Tensor[Float]])]()
+            
+            iterator.foreach{
```

Comment:
It is great that you have used `mapPartitions` as it runs separately on each partition (block) of the RDD, but then after just variable definition, you had to bring the data (from these partitions) to the driver when you apply `iterator.foreach`.  Did you consider taking full advantage of `mapPartitions` here and not doing `foreach`?.